tinygrad/runtime/support/am/amring.py

```diff
@@ -0,0 +1,93 @@
+import os, ctypes
+from tinygrad.runtime.autogen import libpciaccess, amdgpu_2, amdgpu_gc_11_0_0
+from tinygrad.helpers import to_mv, mv_address
+
+class AMRing:
+  def __init__(self, adev, size, pipe, queue, me, vmid, doorbell_index):
+    self.adev, self.size, self.pipe, self.queue, self.me, self.vmid, self.doorbell_index = adev, size, pipe, queue, me, vmid, doorbell_index
+
+    self.mqd_mv = memoryview(bytearray(ctypes.sizeof(amdgpu_2.struct_v11_compute_mqd)))
+    self.mqd = amdgpu_2.struct_v11_compute_mqd.from_address(mv_address(self.mqd_mv))
+
+    self.eop_vm = self.adev.mm.valloc(0x8000)
+    self.mqd_vm = self.adev.mm.valloc(len(self.mqd_mv))
+    self.rptr_vm = self.adev.mm.valloc(0x1000, uncached=True)
+    self.wptr_vm = self.adev.mm.valloc(0x1000, uncached=True)
+    self.ring_vm = self.adev.mm.valloc(self.size, uncached=True)
+
+    self.rptr = self.rptr_vm.cpu_view().cast('Q')
+    self.wptr = self.wptr_vm.cpu_view().cast('Q')
+
+    self.rptr[0] = 0x0
+    self.wptr[0] = 0x0
+
+    self.next_ptr = 0
+
+    self.fill_mqd()
+
+  def fill_mqd(self):
+    self.mqd.header = 0xC0310800
```

Comment:
Does the `ops_amd` driver use the same MQD struct? I see similar stuff on lines 109-122

Reply:
mqd is a ring's context. compute_* are shadowed registers and looks like they are "shadowed" to mqd.

Executing the following in the ring updates its field in the mqd (without any additional syncs/cmds at all). And it inits these registers from mqd as well.
```python
self.q += [amd_gpu.PACKET3(amd_gpu.PACKET3_SET_SH_REG, 1), gfxreg(amd_gpu.regCOMPUTE_STATIC_THREAD_MGMT_SE0), 0xffffffff]
```

MQD also contains HQD regs. It writes them back on a dequeue request into mqd as well. And in `hqd_load` we can load them directly into registers (without any kiq queues now)